### PR TITLE
Feature/flowlet def deprecations

### PR DESCRIPTION
**Pull request by gsps1**
Removes deprecated methods in FlowletDefintion. 
Removing unused transaction classes.
Removing beforeSubmit, onFinish deprecated API for MR, spark.
snapshot codec deprecation removal.

### CODE REVIEW

**Comment by gsps1** [comment]
build running : https://builds.cask.co/browse/CDAP-DUT6191-1

File: cdap-api/src/main/java/co/cask/cdap/api/mapreduce/AbstractMapReduce.java [hunk 1]
```diff
@@ -125,27 +112,18 @@ public final void initialize(MapReduceContext context) throws Exception {
   /**
    * Classes derived from {@link AbstractMapReduce} can override this method to initialize the {@link MapReduce}.
    * {@link MapReduceContext} will be available in this method using {@link AbstractMapReduce#getContext}.
-   * Default implementation of this method calls the deprecated {@link AbstractMapReduce#beforeSubmit} method.
    * @throws Exception if there is any error in initializing the MapReduce
    */
   @TransactionPolicy(TransactionControl.IMPLICIT)
   protected void initialize() throws Exception {
-    beforeSubmit(context);
```
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/mapreduce/AbstractMapReduce.java:133]: I believe it's a warning to have an empty method. Replace these with the 'Do nothing by default' comment

File: cdap-api/src/main/java/co/cask/cdap/api/spark/AbstractSpark.java [hunk 2]
```diff
@@ -147,24 +134,16 @@ public final void initialize(SparkClientContext context) throws Exception {
   /**
    * Classes derived from {@link AbstractSpark} can override this method to initialize the {@link Spark}.
    * {@link SparkClientContext} will be available in this method using {@link AbstractSpark#getContext}.
-   * Default implementation of this method calls the deprecated {@link AbstractSpark#beforeSubmit} method.
    * @throws Exception if there is any error in initializing the Spark
    */
   @TransactionPolicy(TransactionControl.IMPLICIT)
   protected void initialize() throws Exception {
-    beforeSubmit(context);
```
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/spark/AbstractSpark.java:155]: I believe it's a warning to have an empty method. Replace these with the 'Do nothing by default' comment

File: cdap-api/src/main/java/co/cask/cdap/api/mapreduce/AbstractMapReduce.java [hunk 3]
```diff
@@ -125,27 +112,18 @@ public final void initialize(MapReduceContext context) throws Exception {
...
   }
 
   /**
    * Classes derived from {@link AbstractMapReduce} can override this method to destroy the {@link MapReduce}.
-   * Default implementation of this method calls the deprecated {@link AbstractMapReduce#onFinish} method.
    */
   @Override
   @TransactionPolicy(TransactionControl.IMPLICIT)
   public void destroy() {
-    try {
```
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/mapreduce/AbstractMapReduce.java:143]: same here

File: cdap-api/src/main/java/co/cask/cdap/api/spark/AbstractSpark.java [hunk 4]
```diff
@@ -147,24 +134,16 @@ public final void initialize(SparkClientContext context) throws Exception {
...
   }
 
 
   @Override
   @TransactionPolicy(TransactionControl.IMPLICIT)
   public void destroy() {
-    try {
```
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/spark/AbstractSpark.java:162]: same here

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/snapshot/AbstractSnapshotCodec.java [hunk 5]
```diff
@@ -64,8 +64,6 @@ public void encode(OutputStream out, TransactionSnapshot snapshot) {
       encoder.writeLong(snapshot.getTimestamp());
       encoder.writeLong(snapshot.getReadPointer());
       encoder.writeLong(snapshot.getWritePointer());
-      // supporting old versions of codecs
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/snapshot/AbstractSnapshotCodec.java]: do we need this entire class at all? seems like it's unused.

File: cdap-api/src/main/java/co/cask/cdap/api/flow/FlowletDefinition.java [hunk 6]
```diff
@@ -79,9 +74,24 @@ public FlowletDefinition(String flowletName, Flowlet flowlet, int instances) {
                                                        flowletSpec.getDescription(), flowletSpec.getFailurePolicy(),
                                                        flowletSpec.getDataSets(), flowletSpec.getProperties(),
                                                        flowletSpec.getResources());
-    this.streams = Collections.unmodifiableMap(flowletConfigurer.getStreams());
-    this.datasetModules = Collections.unmodifiableMap(flowletConfigurer.getDatasetModules());
-    this.datasetSpecs = Collections.unmodifiableMap(flowletConfigurer.getDatasetSpecs());
+  }
+
+  public FlowletDefinition(String flowletName, Flowlet flowlet,
```
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/flow/FlowletDefinition.java]: do we need both this constructor and the one above? Seems like we should remove the one above and move DefaultFlowletConfigurer out of cdap-api and into cdap-app-fabric.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/flow/DefaultFlowConfigurer.java [hunk 7]
```diff
@@ -80,7 +79,10 @@ public void addFlowlet(String name, Flowlet flowlet) {
   }
 
   private String getFlowletName(Flowlet flowlet) {
-    FlowletDefinition flowletDef = new FlowletDefinition(null, flowlet, 1);
+    DefaultFlowletConfigurer flowletConfigurer = new DefaultFlowletConfigurer(flowlet);
+    flowlet.configure(flowletConfigurer);
+    FlowletSpecification flowletSpecification = flowletConfigurer.createSpecification();
+    FlowletDefinition flowletDef = new FlowletDefinition(name, flowlet, flowletSpecification, 1);
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/flow/DefaultFlowConfigurer.java]: this seems a roundabout way to get flowlet name, but i didn't want to have code duplication, so using this path to get the flowlet name.

**Comment by gsps1** [comment]
@albertshau thanks for the review, addressed the comments

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/flow/DefaultFlowConfigurer.java [hunk 8]
```diff
@@ -78,23 +79,29 @@ public void addFlowlet(String name, Flowlet flowlet) {
   }
 
   private String getFlowletName(Flowlet flowlet) {
-    FlowletDefinition flowletDef = new FlowletDefinition(null, flowlet, 1);
+    DefaultFlowletConfigurer flowletConfigurer = new DefaultFlowletConfigurer(flowlet);
+    flowlet.configure(flowletConfigurer);
+    FlowletSpecification flowletSpecification = flowletConfigurer.createSpecification();
+    FlowletDefinition flowletDef = new FlowletDefinition(null, flowlet, flowletSpecification, 1);
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/flow/DefaultFlowConfigurer.java]: don't think you need this line, can just get it from the flowletSpecification. FlowletDefinition does some field injection, but it doesn't seem like that's needed.

Comment (gsps1) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/flow/DefaultFlowConfigurer.java]: updated it

**Review by albertshau — APPROVED** [review]
lgtm

File: cdap-api/src/main/java/co/cask/cdap/api/flow/FlowletDefinition.java [hunk 9]
```diff
@@ -57,20 +51,14 @@
   private Map<String, Set<Schema>> inputs;
   private Map<String, Set<Schema>> outputs;
 
-  public FlowletDefinition(String flowletName, Flowlet flowlet, int instances) {
-    FlowletSpecification flowletSpec;
-    DefaultFlowletConfigurer flowletConfigurer = new DefaultFlowletConfigurer(flowlet);
-    flowlet.configure(flowletConfigurer);
-    flowletSpec = flowletConfigurer.createSpecification();
-
+  public FlowletDefinition(String flowletName, Flowlet flowlet,
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/flow/FlowletDefinition.java]: It shouldn't need the `Flowlet` anymore

Comment (gsps1) [reply cdap-api/src/main/java/co/cask/cdap/api/flow/FlowletDefinition.java]: updated to pass in `inputTypes` and `outputTypes` map

**Comment by chtyim** [comment]
LGTM

**Comment by gsps1** [comment]
build passed, squashed commits, merging.

**Pull request merged**